File: scripts/train-lora.py
"""
Train LoRA Model for Marshall

This script trains a LoRA model using Replicate's Flux fast finetune.

Prerequisites:
1. Run prepare-lora-training.py first to create the zip file
2. Set REPLICATE_API_TOKEN environment variable

Run with: python scripts/train-lora.py
"""

import os
import replicate
import zipfile
from pathlib import Path

# Configuration
SCRIPT_DIR = Path(__file__).parent
PROJECT_ROOT = SCRIPT_DIR.parent
ZIP_PATH = PROJECT_ROOT / "scripts" / "temp" / "marshall-training-images.zip"
MODEL_NAME = "marshall-lora"
TRIGGER_WORD = "MARSHALL_UNIQ"
TRAINING_STEPS = 1000

# Try to load .env file
try:
    from dotenv import load_dotenv
    env_path = PROJECT_ROOT / ".env.local"
    if env_path.exists():
        load_dotenv(env_path)
    else:
        env_path = PROJECT_ROOT / ".env"
        if env_path.exists():
            load_dotenv(env_path)
except ImportError:
    pass

def main():
    print("🚀 Starting LoRA Training for Marshall\n")
    
    # Check for API token
    api_token = os.environ.get("REPLICATE_API_TOKEN")
    if not api_token:
        print("❌ Error: REPLICATE_API_TOKEN environment variable not set.")
        print("   Set it in your .env.local file or export it in your shell.")
        exit(1)

    # Check if zip file exists
    if not ZIP_PATH.exists():
        print(f"❌ Error: Zip file not found at {ZIP_PATH}")
        print("   Run: python scripts/prepare-lora-training.py first")
        exit(1)

    print("📋 Configuration:")
    print(f"   Model name: {MODEL_NAME}")
    print(f"   Trigger word: {TRIGGER_WORD}")
    print(f"   Training steps: {TRAINING_STEPS}")
    print(f"   Zip file: {ZIP_PATH}\n")

    # Initialize Replicate client
    client = replicate.Client(api_token)
    
    # Get username - try API first, fallback to hardcoded
    username = "jamiesonreinhard2"  # Your Replicate username
    try:
        # Try to get username from API
        import requests
        response = requests.get(
            "https://api.replicate.com/v1/account",
            headers={"Authorization": f"Token {api_token}"}
        )
        if response.ok:
            account_data = response.json()
            username = account_data.get("username", username)
            print(f"✅ Found username: {username}")
    except Exception as e:
        print(f"⚠️  Using hardcoded username: {username}")

    full_model_name = f"{username}/{MODEL_NAME}"
    print(f"📦 Model will be: {full_model_name}\n")

    # Replicate's Flux fast finetune model
    # Try different possible model names/versions
    trainer_models = [
        "black-forest-labs/flux-fast-finetune",
        "black-forest-labs/flux-fast-finetune:latest",
        "black-forest-labs/flux-fast-finetune:main",
    ]
    
    print("📤 Uploading training images and starting training...")
    print("   This may take 1-2 minutes...\n")

    # First, we need to upload the file to Replicate's storage
    # Replicate's Python client can handle this, but we need to use the file path string
    # or upload it first via their API
    
    try:
        # Method 1: Try passing file path as string (Replicate handles upload)
        zip_file_path = str(ZIP_PATH)
        
        # Try each model name until one works
        output = None
        last_error = None
        
        for trainer_model in trainer_models:
            try:
                print(f"   Trying model: {trainer_model}...")
                # Replicate's Python client accepts file paths as strings
                # It will automatically upload them
                output = client.run(
                    trainer_model,
                    input={
                        "input_images": zip_file_path,  # File path - Replicate uploads it
                        "trigger_word": TRIGGER_WORD,
                        "steps": TRAINING_STEPS,
                    }
                )
                print(f"   ✅ Success with model: {trainer_model}\n")
                break
            except Exception as e:
                last_error = e
                if "404" in str(e) or "not found" in str(e).lower():
                    print(f"   ❌ Model not found: {trainer_model}")
                    continue
                else:
                    # Different error, might be file upload issue
                    raise
        
        if output is None:
            raise Exception(f"All model names failed. Last error: {last_error}")

        print("\n✨ Training Complete!\n")
        print("📋 Results:")
        
        # Handle different output formats
        if isinstance(output, str):
            lora_url = output
            print(f"   LoRA Weights URL: {lora_url}")
        elif isinstance(output, list) and len(output) > 0:
            lora_url = output[0] if isinstance(output[0], str) else output[0].get("url", str(output[0]))
            print(f"   LoRA Weights URL: {lora_url}")
        elif isinstance(output, dict):
            lora_url = output.get("lora_weights_url") or output.get("lora_url") or output.get("url")
            print(f"   LoRA Weights URL: {lora_url}")
            if "model_version" in output:
                print(f"   Model Version: {output['model_version']}")
        else:
            print(f"   Output: {output}")
            lora_url = str(output)

        print(f"   Trigger Word: {TRIGGER_WORD}\n")
        
        print("🔧 Next Steps:")
        print("   1. Add to your .env.local:")
        print(f"      MARSHALL_LORA_URL={lora_url}")
        print(f"      MARSHALL_LORA_TRIGGER_WORD={TRIGGER_WORD}")
        print("   2. Set FAL_API_KEY (or use Replicate for inference)")
        print("   3. Start generating images with consistent Marshall face! 🎉\n")

    except Exception as e:
        error_msg = str(e)
        print(f"\n❌ Training failed: {error_msg}")
        
        if "404" in error_msg or "not found" in error_msg.lower():
            print("\n💡 The model might not exist or the name is incorrect.")
            print("   Try checking Replicate's model explorer:")
            print("   https://replicate.com/explore")
            print("\n   Search for 'flux finetune' or 'flux lora training'")
            print("   The correct model might be:")
            print("   - A different organization/user name")
            print("   - A specific version hash needed")
            print("   - Or you might need to use Fal.ai instead (you have credits there)")
        else:
            print("\n💡 Tips:")
            print("   - Check your Replicate account has credits")
            print("   - Verify the zip file is valid")
            print("   - Check Replicate API status")
            print("   - The file might need to be uploaded to a public URL first")
        
        print("\n🔍 Debug info:")
        print(f"   Model tried: {trainer_models}")
        print(f"   Zip file: {ZIP_PATH}")
        print(f"   Zip exists: {ZIP_PATH.exists()}")
        if ZIP_PATH.exists():
            size_mb = ZIP_PATH.stat().st_size / 1024 / 1024
            print(f"   Zip size: {size_mb:.2f} MB")
        
        exit(1)

if __name__ == "__main__":
    main()

File: scripts/train-lora.ts
/**
 * Train LoRA Model for Marshall
 * 
 * This script trains a LoRA model using Replicate's Flux fast finetune.
 * 
 * Prerequisites:
 * 1. Run prepare-lora-training.ts first to create the zip file
 * 2. Upload the zip to a public URL (or use Replicate file upload)
 * 3. Set REPLICATE_API_TOKEN environment variable
 * 
 * Run with: npx tsx scripts/train-lora.ts
 */

import { trainLoRA } from '../lib/ai/lora-training';
import * as fs from 'fs';
import * as path from 'path';

// Configuration
const ZIP_PATH = path.join(__dirname, 'temp', 'marshall-training-images.zip');
const MODEL_NAME = 'marshall-lora';
const TRIGGER_WORD = 'MARSHALL_UNIQ';
const TRAINING_STEPS = 1000;

async function uploadToReplicate(zipPath: string, apiToken: string): Promise<string> {
  console.log('📤 Uploading zip file to Replicate...\n');
  
  const Replicate = (await import('replicate')).default;
  const client = new Replicate({ auth: apiToken });

  try {
    // Replicate can handle file uploads directly
    // We'll use their file upload API or pass the file path
    // For now, we'll need a public URL, so let's check if we can use Replicate's upload
    
    // Read the file
    const fileBuffer = fs.readFileSync(zipPath);
    const fileSizeMB = (fileBuffer.length / 1024 / 1024).toFixed(2);
    
    console.log(`📦 File size: ${fileSizeMB} MB`);
    
    // Replicate's API accepts file uploads via their upload endpoint
    // We'll use the replicate client's file handling
    // For now, we need to upload to a public URL first
    
    // Option 1: Use Replicate's file upload (if supported)
    // Option 2: Upload to a public service and get URL
    
    // For simplicity, let's try using the file directly with Replicate
    // Replicate's Python client handles local files, but TypeScript client might need URL
    
    // We'll need to upload to a public URL first
    // Options: Supabase Storage, S3, or any public file hosting
    
    throw new Error('File upload not yet implemented. Please upload the zip file to a public URL first.');
    
  } catch (error: any) {
    console.error('❌ Upload error:', error.message);
    throw error;
  }
}

async function main() {
  console.log('🚀 Starting LoRA Training for Marshall\n');
  
  // Check for API token
  const apiToken = process.env.REPLICATE_API_TOKEN;
  if (!apiToken) {
    console.error('❌ Error: REPLICATE_API_TOKEN environment variable not set.');
    console.error('   Set it in your .env.local file or export it in your shell.');
    process.exit(1);
  }

  // Check if zip file exists
  if (!fs.existsSync(ZIP_PATH)) {
    console.error(`❌ Error: Zip file not found at ${ZIP_PATH}`);
    console.error('   Run: npx tsx scripts/prepare-lora-training.ts first');
    process.exit(1);
  }

  console.log('📋 Configuration:');
  console.log(`   Model name: ${MODEL_NAME}`);
  console.log(`   Trigger word: ${TRIGGER_WORD}`);
  console.log(`   Training steps: ${TRAINING_STEPS}`);
  console.log(`   Zip file: ${ZIP_PATH}\n`);

  // For now, we need a public URL to the zip file
  // You can upload it to:
  // 1. Supabase Storage (if you have it set up)
  // 2. AWS S3
  // 3. Cloudflare R2
  // 4. Any public file hosting service
  
  console.log('📤 Step 1: Upload zip file to a public URL');
  console.log('   Options:');
  console.log('   - Supabase Storage');
  console.log('   - AWS S3');
  console.log('   - Cloudflare R2');
  console.log('   - Any public file hosting\n');
  
  // Prompt for URL (or we can implement auto-upload)
  const trainingImagesUrl = process.env.TRAINING_IMAGES_URL;
  
  if (!trainingImagesUrl) {
    console.error('❌ Error: TRAINING_IMAGES_URL environment variable not set.');
    console.error('   Please upload the zip file to a public URL and set:');
    console.error('   TRAINING_IMAGES_URL=https://your-public-url.com/marshall-training-images.zip');
    process.exit(1);
  }

  console.log(`✅ Using training images URL: ${trainingImagesUrl}\n`);

  try {
    // Train the LoRA
    console.log('🎓 Starting LoRA training...\n');
    const result = await trainLoRA(apiToken, {
      modelName: MODEL_NAME,
      trainingImagesUrl: trainingImagesUrl,
      triggerWord: TRIGGER_WORD,
      steps: TRAINING_STEPS,
    });

    console.log('\n✨ Training Complete!\n');
    console.log('📋 Results:');
    console.log(`   LoRA Weights URL: ${result.loraWeightsUrl}`);
    console.log(`   Model Version: ${result.modelVersion}`);
    console.log(`   Trigger Word: ${result.triggerWord}\n`);
    
    console.log('🔧 Next Steps:');
    console.log('   1. Add to your .env.local:');
    console.log(`      MARSHALL_LORA_URL=${result.loraWeightsUrl}`);
    console.log(`      MARSHALL_LORA_TRIGGER_WORD=${result.triggerWord}`);
    console.log('   2. Set FAL_API_KEY (or use Replicate for inference)');
    console.log('   3. Start generating images with consistent Marshall face! 🎉\n');

  } catch (error: any) {
    console.error('\n❌ Training failed:', error.message);
    if (error.message.includes('Unexpected')) {
      console.error('\n💡 Tip: Check the Replicate API response format may have changed.');
      console.error('   The training might still be in progress. Check your Replicate dashboard.');
    }
    process.exit(1);
  }
}

main();
